### PR TITLE
Add pep 519

### PR DESCRIPTION
**Pull request by tacaswell**
Thank you so much for your PR! To help us review, fill out the form to the best of your ability. Also, please make use of the development guide at http://matplotlib.org/devdocs/devel/index.html

Replaces  #6788 to support the Path protocol (pep 519)

## PR Checklist
- [x] Has Pytest style unit tests
- [x] Code is PEP 8 compliant 
- [ ] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/whats_new.rst if major new feature
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by anntzer** [comment]
Only quickly skimmed through this, so I may have missed something.
What is the rationale for fspath_no_except, rather than just using (a backport of) fspath?
Of course that'll mean TypeErrors on invalid inputs, but weren't errors going to be raised just after anyways?

**Comment by tacaswell** [comment]
I am not sure, I just rebased @aragilar PR ...

File: lib/matplotlib/backends/backend_agg.py [hunk 1]
```diff
@@ -563,8 +564,9 @@ def print_png(self, filename_or_obj, *args, **kwargs):
             metadata.update(user_metadata)
 
         try:
-            _png.write_png(renderer._renderer, filename_or_obj,
-                           self.figure.dpi, metadata=metadata)
+            _png.write_png(
+                renderer._renderer, fspath_no_except(filename_or_obj),
```
Comment (anntzer) [inline lib/matplotlib/backends/backend_agg.py]: the conversion should almost certainly occur higher, around line 554.

**Comment by aragilar** [comment]
Sorry, this kinda dropped off my todo list... The logic behind fspath_no_except was that there were numerous places where matplotlib allowed either strings/unicode/bytes (which fspath is fine with), or actual files, which fspath would blow up on. fspath_no_except allows them to pass through with no issues. It's written in such a way that when matplotlib drops support for older python versions, the backport of fspath can just be removed without any reworking.

File: lib/matplotlib/backends/backend_agg.py (hunk 1 above)
Comment (aragilar) [inline lib/matplotlib/backends/backend_agg.py]: Yeah, that looks wrong, I don't remember why I did this...

**Comment by tacaswell** [comment]
That makes sense.  This is short enough I would rather vendor this than pick up another tiny dependency.

File: lib/matplotlib/animation.py [hunk 2]
```diff
@@ -349,6 +350,7 @@ def _run(self):
             output = sys.stdout
         else:
             output = subprocess.PIPE
+        command = [fspath_no_except(cmd) for cmd in command]
```
Comment (tacaswell) [inline lib/matplotlib/animation.py]: This should probably just be done on `self.output`

Comment (anntzer) [inline lib/matplotlib/animation.py]: or even better to matplotlib.compat.{Popen,check_output}:
```
subprocess.Popen([Path("ls"), Path("-a"), Path(".")])
```
works (even though this is arguably silly...)

Comment (tacaswell) [inline lib/matplotlib/animation.py]: by 'output', I think I meant 'outfile'..

**Comment by tacaswell** [comment]
super-ceded by #10231 

Thanks for your work on this @aragilar !

**Pull request closed without merging**